akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala

```diff
@@ -484,6 +484,44 @@ object Sink {
     new Sink(scaladsl.Sink.lazyFutureSink { () =>
       create.create().toScala.map(_.asScala)((ExecutionContexts.parasitic))
     }).mapMaterializedValue(_.toJava)
+
+  /**
+   * Handle each stream element with the help of a resource.
+   *
+   * The resource creation function is invoked once when the stream is materialized and the returned resource is passed to
+   * the mapping function for mapping the first element. The mapping function returns a mapped element to emit
+   * downstream. The returned `T` MUST NOT be `null` as it is illegal as stream element - according to the Reactive Streams specification.
+   *
+   * The `close` function is called only once when the upstream or downstream finishes or fails. You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Early completion can be done with combination of the [[takeWhile]] operator.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * You can configure the default dispatcher for this Source by changing the `akka.stream.materializer.blocking-io-dispatcher` or
+   * set it for a given Source by using [[ActorAttributes]].
+   *
+   * '''Backpressures when''' when the previous fold function invocation has not yet completed
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' never
+   *
+   * @tparam R the type of the resource
+   * @tparam T the type of the output elements
+   * @param create function that creates the resource
+   * @param f function that handles the upstream element with resource.
+   * @param close function that closes the resource
+   */
+  def foldResource[R, T](
+      create: Creator[R],
+      f: function.Procedure2[R, T],
+      close: function.Procedure[R]): Sink[T, CompletionStage[Done]] =
```

Comment:
Can we switch to the `java.util.function.*`?

Reply:
We use our own where we want to allow throwing checked exceptions, like in actor message handling where it makes sense, but here the JDK ones makes more sense indeed.